### ISSUE TITLE
%placeholder bugs parent (&) selector when placed inside a css selector

### ISSUE THREAD

**Issue by RaphaelDDL** [sass/sass#1885]
Hi there,

I found out a bug that when you put your `%placeholder` inside a css selector, parent selector `&` duplicates the parent selector. I feel is a bug for both compiling wrongly because of a non-rendering class and because there's no restriction stated @ [sass](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#placeholder_selectors_) [docs](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#placeholders) for where to place them.

Given this following example:

```
%brandColor {
    color: red;
}

body {
    color: black;

    &.brandColor {
        @extend %brandColor;
    }
}
```

That compiles correctly into:

```
body.brandColor {
    color: red;
}
body {
    color: black;
}
```

[jsFiddle for correct compile](http://jsfiddle.net/RaphaelDDL/woqoh180/)
## 

But when the `%placeholder` is placed inside the selector (`body` in this case):

```
body {
    color: black;
    %innerColor {
        color: green;
    }

    &.innerColor {
        @extend %innerColor;
    }
}
```

The result is that & duplicates the parent selector:

```
body {
    color: black;
}
body body.innerColor {
    color: green;
}
```

[jsFiddle for incorrect compile](http://jsfiddle.net/RaphaelDDL/woqoh180/1/)
## 

This happens to any selector (element, id, class). The 'fix' as for now is to keep placeholders outside selectors, but I wanted to put inside of parent selector as the placeholder would be `@extend`ed exclusively by classes inside that parent (`body` in this example).

Reproducible in latest stable: `Sass 3.4.19 (Selective Steve)` with `sass test.scss test.css`.
Was reproducible on 3.2.x too (the version I was when I noticed this).
## 

**Edit:**

Also another issue with placeholders. Considering the first example, where the `&.brandColor` is an inner rule, shouldn't the compiled version be _after_ the `body` one, not before?

See [this jsFiddle](http://jsfiddle.net/RaphaelDDL/woqoh180/2/). If the placeholder is after, then it compiles following the correct flow (parent then inner rules).
## 

Best Regards,
RaphaelDDL

**Comment by chriseppstein** [comment]
This is not a bug. It is working as intended.

>  I wanted to put inside of parent selector as the placeholder would be @extended exclusively by classes inside that parent

Keeping the base selector outside of the scoped selector is the right thing. You accomplish your goal by only extending the base selector in your scoped context.

**Comment by RaphaelDDL** [comment]
Oh, My bad @chriseppstein :'(
Seems there was one issue for that already (#1172). Guess I missed it because I just searched for 'placeholder', not 'placeholder parent' to filter out better.

May I suggest improving the doc to mention that `%placeholder` has scope if used inside a selector?

Thank you.

**Comment by chriseppstein** [comment]
@RaphaelDDL You read this documentation? http://sass-lang.com/documentation/file.SASS_REFERENCE.html#placeholders It seems to cover it from my understanding of it, but maybe I'm too close to the problem.

**Comment by RaphaelDDL** [comment]
@chriseppstein Yup, I did (I even linked it on first post).

I've re-read it over and over but I couldn't make the link that `#id { %placeholder {} }` would be the same as `#id %placeholder {}`. For me, they are two different scenarios from what I would've call a 'placeholder for replace' and a 'truly placeholder' (I'll explain below). 

Looking at [the example](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#placeholders), it shows an use that it works as a "placeholder for replace" for the class you are extending it from. So if I do the placeholder as part of a CSS selector (say `#id %placeholder {}`), then when I do `.whatever { @extend %placeholder; }`, I should expect my class to "replace" the placeholder part, the same way it would work with regex `()` matchers which then you replace/reefer the parts with `$n`. Thus resulting `#id .whatever {}`.

But I thought that, without being part of a CSS selector - simply a "truly placeholder" class -, wouldn't affect the outcome. [Here](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#placeholder_selectors_) there's this quote:

> On their own, without any use of @extend, rulesets that use placeholder selectors will not be rendered to CSS.

From that I assumed they wouldn't do anything nor should have selector "scope" because they are "on their own", that is, simply `%placeholder` independent of position.

But I could truly understand after reading nex3's explanation in #1172. I realized that what I thought "without being part of a CSS selector" is utterly wrong. When we do 
`#id { %placeholder {} }` we are indeed making it as part of the CSS selector, using the nesting provided by sass engine.. Therefore it is indeed the same as `#id %placeholder {}`.

Not sure if it's my own problem, since my native language isn't English. But I feel (less) dumb knowing I wasn't the only person that thought the way I thought lol. Anyway, deeply sorry if I didn't make any sense..

Thank you for your time reading through this.
And thank you again, this time for Compass lol :)
